lsp-mode.el

```diff
@@ -1968,7 +1983,96 @@ WORKSPACE is the workspace that contains the progress token."
+                      (arrow-icon (lsp--headerline-breadcrumb-arrow-icon)))
+                  (format "%s %s %s"
+                          last-symbol-name
+                          arrow-icon
+                          (if symbol2-icon
+                              (concat symbol2-icon symbol2-name)
+                            symbol2-name))))
+              symbols-hierarchy ""))
+
+(defun lsp--headerline-document-symbols->symbols-hierarchy (document-symbols)
+  "Convert DOCUMENT-SYMBOLS to symbols hierarchy."
+  (-let (((symbol &as &DocumentSymbol? :children?) (seq-some (-lambda ((symbol &as &DocumentSymbol :range))
+                                                          (-let (((beg . end) (lsp--range-to-region range)))
+                                                            (and (<= beg (point) end)
+                                                                 symbol)))
+                                                        document-symbols)))
+    (if children?
+        (append (list symbol) (lsp--headerline-document-symbols->symbols-hierarchy children?))
+      (when symbol
+        (list symbol)))))
+
+(defun lsp--headerline-symbols-informations->symbols-hierarchy (symbols-informations)
+  "Convert SYMBOL-INFORMATIONS to symbols hierarchy."
+  (seq-some (-lambda ((symbol &as &SymbolInformation :location (&Location :range)))
+              (-let (((beg . end) (lsp--range-to-region range)))
+                (and (<= beg (point) end)
+                     symbol)))
+            symbols-informations))
+
+(defun lsp--headerline-symbols->symbols-hierarchy (symbols)
+  "Convert SYMBOLS to symbols-hierarchy."
+  (when-let (first-symbol (lsp-seq-first symbols))
+    (if (lsp-symbol-information? first-symbol)
+        (lsp--headerline-symbols-informations->symbols-hierarchy symbols)
+      (lsp--headerline-document-symbols->symbols-hierarchy symbols))))
+
+(defun lsp--headerline-check-breadcrumb (&rest _)
+  "Request for document symbols to build the breadcrumb."
+  (when (lsp-feature? "textDocument/documentSymbol")
+    (-if-let* ((lsp--document-symbols-request-async t)
+               (symbols (lsp--get-document-symbols))
+               (symbols-hierarchy (lsp--headerline-symbols->symbols-hierarchy symbols)))
+        (setq lsp--headerline-breadcrumb-string (lsp--headerline-build-string symbols-hierarchy))
+      (setq lsp--headerline-breadcrumb-string nil))))
+
+(define-minor-mode lsp-headerline-breadcrumb-mode
+  "Toggle breadcrumb on headerline."
+  :group 'lsp-mode
+  :global nil
+  (cond
+   (lsp-headerline-breadcrumb-mode
+    (add-to-list 'header-line-format '(t (:eval lsp--headerline-breadcrumb-string)))
+    (add-hook 'lsp-on-idle-hook 'lsp--headerline-check-breadcrumb nil t))
+   (t
+    (remove-hook 'lsp-on-idle-hook 'lsp--headerline-check-breadcrumb t)
+    (setq-local header-line-format (remove '(t (:eval lsp--headerline-breadcrumb-string)) header-line-format)))))
```

Comment:
It seems that you're adding to (potentially) global `header-line-format` but removing from local `header-line-format`.
I think you can just use `setq` here. It has similar effect if `header-line-format` is already buffer local, and working even if there's only one global `header-line-format`